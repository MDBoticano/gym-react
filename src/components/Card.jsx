import React, { useContext } from 'react';
import styled, { ThemeContext } from 'styled-components';

import TagsRow from './TagsRow';

const StyledCard = styled.div`
  margin: 0.5rem 0;
  padding: 1rem;
  border-radius: 0.5rem;

  background-color: white;
  box-shadow: 1px 1px 4px rgba(0, 0, 0, 0.5);

  &:hover {
    box-shadow: 1px 1px 2px rgba(0, 0, 0, 0.25);
  }
`;

const CardTitle = styled.p`
  margin: 0;

  font-family: ${props => props.theme.fonts.standard};
  font-size: ${props => props.theme.fontSize.xl};
  line-height: 1.75rem;
`;

const CardDescription = styled.p`
  margin: 0;
  margin-top: 5px;

  max-height: 2.5rem;
  overflow: hidden;
  text-overflow: ellipsis;

  font-family: ${props => props.theme.fonts.standard};
  font-size: ${props => props.theme.fontSize.m};
  line-height: 1.25rem;

  color: gray;
`;

const Card = (props) => {
  const theme = useContext(ThemeContext);
  return (
    <StyledCard theme={theme}>
      <CardTitle>{props.title}</CardTitle>
      <CardDescription>{props.description}</CardDescription>
      <TagsRow tags={props.tags} />
    </StyledCard>
  );
};

export default Card;
