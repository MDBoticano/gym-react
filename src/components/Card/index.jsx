import React, { useState, useContext } from 'react';
import { ThemeContext } from 'styled-components';

import {
  StyledCard,
  CardTitle,
  CardToggle,
  CardDescription,
  CardActions,
  DeleteButton
} from './style';

import { TagsRow } from '../TagsRow';

export const Card = (props) => {
  const [collapsed, setCollapsed] = useState(true);

  const collapse = () => setCollapsed(!collapsed);

  const theme = useContext(ThemeContext);
  const { id, name, description, tags } = props.exercise;
  const deleteExercise = props.deleteExercise;
  return (
    <StyledCard theme={theme} collapsed={collapsed}>
      <div className="card-grid">
        <CardTitle>{name}</CardTitle>
        <CardToggle collapsed={collapsed} onClick={() => collapse()} />
      </div>
      <CardDescription collapsed={collapsed}>
        {description}
      </CardDescription>
      <CardActions>
        <TagsRow tags={tags} collapsed={collapsed}/>
        <DeleteButton onClick={() => deleteExercise(id)} />
      </CardActions>
    </StyledCard>
  );
};
