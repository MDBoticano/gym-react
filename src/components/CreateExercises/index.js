import React, { useState } from 'react';

import {
  StyledForm,
  StyledFormContainer,
  StyledCallToAction,
} from './style';

const CallToAction = ({ callback }) => {
  return (
    <StyledCallToAction onClick={() => callback()}>
      Create an Exercise
    </StyledCallToAction>
  );
}

export const CreateExercises = ({ addExercise }) => {
  const [hidden, setHidden] = useState(false);

  const [formInputs, setFormInputs] = useState({
    name: "",
    description: "",
    tags: []
  });

  const handleFormChange = (fieldChanged, event) => {
    const newValue = event.target.value;
    const newInputs = {
      ...formInputs,
      [fieldChanged] : newValue
    };

    setFormInputs(newInputs);
  }

  const toggleHidden = () => setHidden(!hidden);

  const handleFormSubmit = (event) => {
    event.preventDefault();
    addExercise(formInputs);
    toggleHidden();
  };

  if (hidden) { return <CallToAction callback={toggleHidden} /> }

  return (
    <StyledFormContainer>
      <StyledForm onSubmit={handleFormSubmit}>
        <label>
          Name:
          <input 
            type="text"
            value={formInputs.name}
            onChange={(event) => handleFormChange('name', event)}
          />
        </label>
        <label>
          Description:
          <input 
            type="text"
            value={formInputs.description}
            onChange={(event) => handleFormChange('description', event)}
          />
        </label>
        <input type="submit" value="Submit" />
      </StyledForm>
    </StyledFormContainer>
  );
};
