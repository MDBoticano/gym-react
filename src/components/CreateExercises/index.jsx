import React, { useState } from 'react';

import {
  StyledLabel,
  StyledInput,
  StyledTextArea,
  StyledSubmitForm,
  StyledCancelForm,
  StyledFormEntry,
  StyledFormControls,
  StyledForm,
  StyledFormContainer,
  StyledCallToAction,
  StyledTagDefault,
  StyledTagActive,
} from './style';

import { DummyTags } from '../../data/DummyTags';

const CallToAction = ({ callback }) => {
  return (
    <StyledCallToAction onClick={() => callback()}>
      Create an Exercise
    </StyledCallToAction>
  );
}

const initialFormState = { name: "", description: "", tags: [] };

export const CreateExercises = ({ addExercise }) => {
  const [hidden, setHidden] = useState(true);

  const [formInputs, setFormInputs] = useState({...initialFormState});

  const handleFormChange = (fieldChanged, event) => {
    const newValue = event.target.value;
    const newInputs = {
      ...formInputs,
      [fieldChanged] : newValue
    };

    setFormInputs(newInputs);
  }

  const toggleHidden = () => setHidden(!hidden);

  const resetForm = () => {
    const newFormInputs = {...initialFormState};
    newFormInputs.tags = [];
    setFormInputs(newFormInputs);
  }

  

  const closeForm = (event) => {
    event.preventDefault();
    toggleHidden();
    resetForm();
  }

  const closeFormModal = (event) => {
    const target = event.target;
    console.log(event.type);

    // Close method 1: click outside the modal on the container (background)
    if (event.type === "click" && target.className.includes("modal-overlay")) {
      closeForm(event);
    }
    // Close method 2: pressing escape
    // TODO: doesn't work until the form is focused
    if (event.type === "keydown" && event.keyCode === 27) {
      closeForm(event);
    }
  }

  const handleFormSubmit = (event) => {
    event.preventDefault();
    addExercise(formInputs);
    toggleHidden();
    resetForm();
  };

  /* Tags Adding Handler */
  const AddTags = ({ formInputs }) => {
    const allTags = DummyTags; 

    const addTagToState = (tag) => {
      const newInputs = {...formInputs};
      newInputs.tags.push(tag);
      setFormInputs(newInputs);
    }

    const removeTagFromState = (tagToRemove) => {
      const newInputs = {...formInputs};
      newInputs.tags = newInputs.tags.filter(tag => tag !== tagToRemove);
      setFormInputs(newInputs);
    }

    const makeTagComponents = (tagsStrings) => {
      const tagComponents = tagsStrings.map((tag) => {
        const isSelected = formInputs.tags.includes(tag);

        const tagButton = isSelected ? 
        <StyledTagActive onClick={() => removeTagFromState(tag)} key={tag}>
            {tag}
        </StyledTagActive> :
        <StyledTagDefault onClick={() => addTagToState(tag)} key={tag}>
            {tag}
        </StyledTagDefault>

        return tagButton;
      });

      return tagComponents;
    }
    
    return (
      <div className="AddTags">
        {makeTagComponents(allTags)}
      </div>
    );
  }
  
  if (hidden) { return <CallToAction callback={toggleHidden} /> }

  return (
    <StyledFormContainer className="modal-overlay" tabIndex="0"
      onClick={(event) => closeFormModal(event)}
      onKeyDown={(event) => closeFormModal(event)}
    >
      <StyledForm>
        <StyledFormEntry>
          <StyledLabel htmlFor="name">Name</StyledLabel>
          <StyledInput 
            name="name"
            type="text"
            value={formInputs.name}
            onChange={(event) => handleFormChange('name', event)}
          />
        </StyledFormEntry>

        <StyledFormEntry>
          <StyledLabel htmlFor="description">Description</StyledLabel>
          <StyledTextArea
            name="description"
            type="textarea"
            value={formInputs.description}
            onChange={(event) => handleFormChange('description', event)}
          />
        </StyledFormEntry>

        <StyledFormEntry>
          <StyledLabel htmlFor="tags">Tags</StyledLabel>
          <AddTags formInputs={formInputs} />
        </StyledFormEntry>
       
        <StyledFormControls>
          <StyledCancelForm onClick={(event) => closeForm(event)}>
            cancel
          </StyledCancelForm>
          <StyledSubmitForm
            type="submit"
            onClick={(event) => handleFormSubmit(event)}
          >
            Submit
          </StyledSubmitForm>
        </StyledFormControls>
      </StyledForm>
    </StyledFormContainer>
  );
};
