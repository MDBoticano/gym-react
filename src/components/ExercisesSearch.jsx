import React, { useState } from 'react';

import styled from 'styled-components';

import debounce from 'lodash.debounce'; //implement yourself later

import { filterData } from '../utilities/search-utilities';


const StyledSearch = styled.div`
  margin: 0;
  padding: 2rem 1rem 1rem;

  background-color: white;
`;

// Debounce delay in miliseconds
const DEBOUNCE_TIME = 900;

const debouncedFilter = debounce((
  data, queryValue, queryables, setActiveData
) => filterData(data, queryValue, queryables, setActiveData), DEBOUNCE_TIME);

const ExercisesSearch = ({ data, setActiveData }) => {

  const [query, setQuery] = useState('');

  const [queryables] = useState({
    name: true,
    description: false,
    tags: true,
  });

  const handleChange = (event) => {
    const queryValue = event.target.value;

    setQuery(queryValue);
    debouncedFilter(data, queryValue, queryables, setActiveData);
  }

  return (
    <StyledSearch className="ExercisesSearch">
       <input
        type="text"
        value={query}
        placeholder={"search"}
        onChange={(e) => handleChange(e)}
      />
    </StyledSearch>
  );
};

export default ExercisesSearch;
