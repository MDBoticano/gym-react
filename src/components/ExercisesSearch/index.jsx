import React, { useContext } from 'react';
import { ThemeContext } from 'styled-components';

import { StyledSearch, StyledInput } from './style';

export const ExercisesSearch = ({ query, setQuery }) => {
  const theme = useContext(ThemeContext);

  return (
    <StyledSearch className="ExercisesSearch" theme={theme}>
       <StyledInput
        type="text"
        value={query}
        placeholder={"Search"}
        onChange={(event) => setQuery(event.target.value)}
      />
    </StyledSearch>
  );
};
