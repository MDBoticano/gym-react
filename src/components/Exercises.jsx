import React, { useState, useEffect, useContext } from 'react';

import ExercisesSearch from './ExercisesSearch';
import { CreateExercises } from './CreateExercises';
import Card from './Card';
import { displayMessage } from './Messages';


import DummyExercises from '../data/DummyExercises';

import debounce from 'lodash.debounce'; //implement yourself later
import { filterData } from '../utilities/search-utilities';

import { SearchProvider } from './SearchContext';
import styled, { ThemeContext } from 'styled-components';


// Debounce delay in miliseconds
const DEBOUNCE_TIME = 500;
const DEBOUNCE_OPTIONS = {};
const debouncedFilter = debounce((...args) => filterData(...args),
  DEBOUNCE_TIME, DEBOUNCE_OPTIONS
);

const Exercises = () => {
  const [message, setMessage] = useState({ 
    message: "", 
    messageType: "" 
  });
  const [exercises, setExercises] = useState(DummyExercises);
  const [activeExercises, setActiveExercises] = useState([]);
  const [queryText, setQueryText] = useState('');
  const [queryables] = useState({
    name: true,
    description: false,
    tags: true,
  });

  useEffect(() => {
    debouncedFilter(exercises, queryText, queryables, setActiveExercises);
  }, [exercises, queryText, queryables]);

  const displayMessageTemporarily = (message, messageType, time = 3000) => {
    setMessage({ message: message, messageType: messageType});

    setTimeout(() => {
      setMessage({ message: "", messageType: ""})
    }, time);
  }

  const addExercise = (exercise) => {
    const tempId = exercises.length + 1;
    exercise.id = tempId;

    if (exercise.name === "") {
      displayMessageTemporarily( "Your exercise needs a name!", "error");
      return;
    }

    const newExercises = [exercise, ...exercises];
    displayMessageTemporarily(`Added ${exercise.name}`, "success");


    setExercises(newExercises);
  }

  const cardsList = (allExercises) => {
    const loadingCopy = "Retrieving exercises...";
    const noResultsCopy = "We couldn't find what you were looking for. Try refining your search."

    // Deal with empty cases: while loading & when a search returns nothing
    if (allExercises.length === 0) {
      const makeEmptyResultMessage = (query) => {
        return query === '' ? loadingCopy : noResultsCopy;
      }
      return (
        <StyledEmptyResult>
          {makeEmptyResultMessage(queryText)}
        </StyledEmptyResult>
      );
    } else {
      const exerciseCards = allExercises.map((exercise) => (
        <Card className="Card" exercise={exercise} key={exercise.id} />
      ));

      return exerciseCards;
    }
  }

  const theme = useContext(ThemeContext);
  const searchContext = { setQuery: setQueryText };

  return (
    <SearchProvider value={searchContext}>
      <StyledExercises theme={theme}>
        <StyledHeader>
          <p className="page-title">Exercises</p>
          <ExercisesSearch query={queryText} setQuery={setQueryText} />
          {displayMessage(message.message, message.messageType)}
        </StyledHeader>
        <StyledList className="ExercisesList">
          {cardsList(activeExercises)}
        </StyledList>
        <CreateExercises addExercise={addExercise} />
      </StyledExercises>
    </SearchProvider>
  );
};

const StyledExercises = styled.div`
  height: 100vh;

  margin: 0;
  padding: 0;

  display: flex;
  flex-direction: column;
`;

const StyledHeader = styled.header`
  margin: 0;
  padding-top: 1.5rem;

  background-color: white;

  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.5);

  /* so cards are below the header despite sibling */
  z-index: 100; 

  .page-title {
    margin: 0;
    padding: 0;

    padding-left: 1rem;

    font-family: ${props => props.theme.fonts.standard};
    font-size: ${props => props.theme.fontSize.xxl};
    line-height: 2.5rem;
  }
`;

const StyledList = styled.div`
  margin: 0;
  margin-bottom: 4rem; /* height of add exercise button */
  padding: 0 0.5rem;

  overflow: auto;
`;

const StyledEmptyResult = styled.div`
  margin: 0.5rem 0;
  padding: 1rem;

  color: hsl(0, 0%, 29%);

  font-family: ${props => props.theme.fonts.standard};
  font-size: ${props => props.theme.fontSize.l};

  text-align: center;
`;

export default Exercises;
