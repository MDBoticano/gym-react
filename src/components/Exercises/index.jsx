import React, { useState, useEffect, useContext } from 'react';
import { ThemeContext } from 'styled-components';
import { SearchProvider } from '../SearchContext';

import debounce from 'lodash.debounce'; //implement yourself later
import { filterData } from '../../utilities/search-utilities';

import gymServices from '../../services/gymServices';

import { ExercisesSearch } from '../ExercisesSearch';
import { CreateExercises } from '../CreateExercises';
import { Card } from '../Card';
import { displayMessage } from '../Messages';

import DummyExercises from '../../data/DummyExercises';

import { 
  StyledExercises,
  StyledHeader, 
  StyledList, 
  StyledEmptyResult 
} from './style';

// Debounce delay in miliseconds
const DEBOUNCE_TIME = 500;
const DEBOUNCE_OPTIONS = {};
const debouncedFilter = debounce((...args) => filterData(...args),
  DEBOUNCE_TIME, DEBOUNCE_OPTIONS
);    

export const Exercises = () => {
  const [message, setMessage] = useState({ 
    message: "", 
    messageType: "" 
  });
  const [exercises, setExercises] = useState(DummyExercises);
  const [activeExercises, setActiveExercises] = useState([]);
  const [queryText, setQueryText] = useState('');
  const [queryables] = useState({
    name: true,
    description: false,
    tags: true,
  });

  useEffect(() => {
    const asyncGetExercises = async() => { 
      const dbExercises = await gymServices.getExercises();
      // console.log(dbExercises);

      // Fallback for when the backend has no data
      dbExercises.length > 0
      ? setExercises(dbExercises)
      : setExercises(DummyExercises)
    };

    asyncGetExercises();
  }, []);

  useEffect(() => {
    debouncedFilter(exercises, queryText, queryables, setActiveExercises);
  }, [exercises, queryText, queryables]);

  const displayMessageTemporarily = (message, messageType, time = 3000) => {
    setMessage({ message: message, messageType: messageType});

    setTimeout(() => {
      setMessage({ message: "", messageType: ""})
    }, time);
  }

  const addExercise = (exercise) => {
    if (!exercise.id) { 
      const tempId = exercises.length + 1;
      exercise.id = tempId;
    }

    if (exercise.name === "") {
      displayMessageTemporarily( "Your exercise needs a name!", "error");
      return;
    }

    const newExercises = [exercise, ...exercises];
    displayMessageTemporarily(`Added ${exercise.name}`, "success");


    setExercises(newExercises);
  }

  const deleteExercise = async (exerciseId) => {
    const updatedExercises = await gymServices.deleteExercise(exerciseId);
    console.log(updatedExercises);
    setExercises(updatedExercises);
  }

  const cardsList = (allExercises) => {
    const loadingCopy = "Retrieving exercises...";
    const noResultsCopy = "We couldn't find what you were looking for. Try refining your search."

    // Deal with empty cases: while loading & when a search returns nothing
    if (!allExercises || allExercises.length === 0) {
      const makeEmptyResultMessage = (query) => {
        return query === '' ? loadingCopy : noResultsCopy;
      }
      return (
        <StyledEmptyResult>
          {makeEmptyResultMessage(queryText)}
        </StyledEmptyResult>
      );
    } else {
      const exerciseCards = allExercises.map((exercise) => (
        <Card 
          exercise={exercise} 
          deleteExercise={deleteExercise}
          key={exercise.id} 
        />
      ));

      return exerciseCards;
    }
  }

  const theme = useContext(ThemeContext);
  const searchContext = { setQuery: setQueryText };

  return (
    <SearchProvider value={searchContext}>
      <StyledExercises theme={theme}>
        <StyledHeader>
          <p className="page-title">Exercises</p>
          <ExercisesSearch query={queryText} setQuery={setQueryText} />
          {displayMessage(message.message, message.messageType)}
        </StyledHeader>
        <StyledList className="ExercisesList">
          {cardsList(activeExercises)}
        </StyledList>
        <CreateExercises addExercise={addExercise} />
      </StyledExercises>
    </SearchProvider>
  );
};
