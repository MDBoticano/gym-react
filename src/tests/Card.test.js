import React from 'react';
import { Card } from '../components/Card';
import ThemeContext from '../ThemeContext';

import { create } from 'react-test-renderer';
import 'jest-styled-components';

const sampleExercise = {
  name: 'Squat',
  description: 'A squat is a strength exercise in which the trainee lowers their hips from a standing position and then stands back up.',
  tags: [ 'Quads', 'Glutes', 'Abs' ],
  id: 1,
}

test('Card', () => {
  const component = create(
    <ThemeContext>
      <Card exercise={sampleExercise} />
    </ThemeContext>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});