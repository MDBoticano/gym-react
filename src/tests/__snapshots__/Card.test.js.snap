// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card 1`] = `
.c0 {
  margin: 0.5rem 0;
  padding: 1rem;
  border-radius: 0.5rem;
  background-color: white;
  box-shadow: 1px 1px 4px rgba(0,0,0,0.5);
}

.c0:hover {
  box-shadow: 1px 1px 2px rgba(0,0,0,0.25);
}

.c0 .card-grid {
  display: grid;
  grid-template-columns: 1fr 1.5rem;
}

.c1 {
  margin: 0;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','Roboto','Oxygen','Ubuntu','Helvetica Neue',Arial,sans-serif;
  font-size: 1.500rem;
  line-height: 1.75rem;
}

.c3 {
  visibility: hidden;
  margin: 0 0;
  padding: 0;
  max-height: 0;
  overflow: hidden;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','Roboto','Oxygen','Ubuntu','Helvetica Neue',Arial,sans-serif;
  font-size: 1.000rem;
  line-height: 1.25rem;
  color: gray;
}

.c2 {
  margin: 0.375rem auto;
  padding: 0;
  height: 1rem;
  width: 1rem;
  background-image: url('images/Expand.svg');
  background-size: cover;
}

.c4 {
  margin: 0;
  padding: 0;
  display: grid;
  grid-template-columns: 1fr 1.5rem 1.5rem;
  grid-gap: 0.5rem;
}

.c8 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0.25rem;
  border: none;
  background: url('images/Trash.svg') center no-repeat;
  background-size: contain;
}

.c8:hover {
  cursor: pointer;
}

.c6 {
  height: 1.75rem;
  margin: 0.25rem 0 0;
  padding: 0 0.5rem;
  border: solid 1px hsl(0,0%,59%);
  border-radius: 0.25rem;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','Roboto','Oxygen','Ubuntu','Helvetica Neue',Arial,sans-serif;
  font-size: 0.750rem;
  line-height: 1.25rem;
  color: hsl(0,0%,59%);
  background-color: white;
  outline: none;
}

.c6:not(:last-child) {
  margin-right: 0.25rem;
}

.c6:hover {
  border: solid 1px hsl(0,0%,39%);
  color: white;
  background-color: hsl(0,0%,39%);
}

.c5 {
  height: 2rem;
  overflow: hidden;
}

.c7 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0.25rem;
  border: none;
  background: url('images/Edit.svg') center no-repeat;
  background-size: contain;
}

.c7:hover {
  cursor: pointer;
}

<div
  className="c0"
>
  <div
    className="card-grid"
  >
    <p
      className="c1"
    >
      Squat
    </p>
    <div
      className="c2"
      onClick={[Function]}
    />
  </div>
  <p
    className="c3"
  >
    A squat is a strength exercise in which the trainee lowers their hips from a standing position and then stands back up.
  </p>
  <div
    className="c4"
  >
    <div
      className="c5 tags-row"
    >
      <button
        className="c6 Tag"
        onClick={[Function]}
      >
        Quads
      </button>
      <button
        className="c6 Tag"
        onClick={[Function]}
      >
        Glutes
      </button>
      <button
        className="c6 Tag"
        onClick={[Function]}
      >
        Abs
      </button>
    </div>
    <div
      className="c7"
      onClick={[Function]}
    />
    <div
      className="c8"
      onClick={[Function]}
    />
  </div>
</div>
`;
